tools/c7n_azure/c7n_azure/resources/storage.py

```diff
@@ -104,3 +115,131 @@ def _query_rules(self, resource):
         resource_rules = set([IPNetwork(r['value']) for r in ip_rules])
 
         return resource_rules
+
+
+@Storage.filter_registry.register('storage-diagnostic-settings')
+class StorageDiagnosticSettingsFilter(ValueFilter):
+    BLOB_TYPE = 'blob'
+    QUEUE_TYPE = 'queue'
+    TABLE_TYPE = 'table'
+    FILE_TYPE = 'file'
+
+    def __init__(self, data, manager=None):
+        super(StorageDiagnosticSettingsFilter, self).__init__(data, manager)
+        self.storage_type = data['storage_type']
+
+    schema = type_schema('storage-diagnostic-settings',
+                         rinherit=ValueFilter.schema,
+                         storage_type={
+                             'type': 'string',
+                             'enum': [BLOB_TYPE, QUEUE_TYPE, TABLE_TYPE, FILE_TYPE]},
+                         required=['storage_type'],
+                         )
+
+    def process(self, resources, event=None):
+        futures = []
+        results = []
+        session = local_session(self.manager.session_factory)
+        # Process each resource in a separate thread, returning all that pass filter
+        with self.executor_factory(max_workers=3) as w:
```

Comment:
Can we use `ThreadHelper.execute_in_parallel` here?

Reply:
I had to remind myself why I couldn't use it. It's because I need to pass the session through. I can try and be smart about it, but then it breaks the pattern of not setting class vars out of the init method. How much do we care about that style guideline?

Reply:
I could also extend the ThreadHelper method to take additional parameters.